Docker container Ubuntu 14.04/Erlang 18.2 for build.a.o

basti1302 opened the pull request:
This PR adds a Dockerfile which will produce a container based on Ubuntu 14.04 LTS with Erlang 18.2 and all required dependencies to build current CouchDB master as discussed in [INFRA-10126](https://issues.apache.org/jira/browse/INFRA-10126)).

It also removes a lot of the prototypical stuff that is not needed anymore (based on the discussion in the ticket mentioned above), like provisioning a Jenkins master, configuring build jobs via scm-sync-plug-in etc.

kxepal commented on ansible/roles/ubuntu-worker/tasks/dependencies.yml:
```diff
@@ -0,0 +1,59 @@
+  - texinfo
+
+# More packages found in klaemo's Dockerfile... not sure if required, apparently not?
+# - erlang-base-hipe
```
We can live without hipe, for sure.

kxepal commented on ansible/roles/ubuntu-worker/tasks/dependencies.yml:
```diff
@@ -0,0 +1,59 @@
+
+# More packages found in klaemo's Dockerfile... not sure if required, apparently not?
+# - erlang-base-hipe
+# - erlang-dialyzer
```
Dialyzer is need for more complex code inspect which cost significant amount of memory and CPU resources. Would be nice to run it on CI, but it will require some special policy.

kxepal commented on ansible/roles/ubuntu-worker/tasks/dependencies.yml:
```diff
@@ -0,0 +1,59 @@
+# More packages found in klaemo's Dockerfile... not sure if required, apparently not?
+# - erlang-base-hipe
+# - erlang-dialyzer
+# - erlang-eunit
```
This requires for tests suite run.

basti1302 replied:
Still, a full `make` runs fine without erlang-eunit. I guess the unit tests should definitely run on CI. Do I need to call anything special to trigger them? Or does the Makefile simply skip them if erlang-eunit is not installed?

kxepal replied:
Just `make` build the sources - it hardly can trigger tests run. `make check` runs all the tests. `make eunit` runs only eunit tests. `make javascript` - only javascript integration test suite. And so on.

basti1302 replied:
Soo, which steps exactly would you except to run in a regular CI build. `make && make check`?

kxepal replied:
Actually, `make && make distcheck` (oops, I just realized that we don't have `distcheck`, so `make all check dist`) - that we build and right, all tests passed and we're able to produce ready-to-ship tarball.

kxepal commented on ansible/roles/ubuntu-worker/tasks/dependencies.yml:
```diff
@@ -0,0 +1,59 @@
+# - libnspr4
+# - libnspr4-0d
+# - libnspr4-dev
+# - openssl
```
IIRC, openssl requires for Erlang SSL correct work.

rnewson replied:
yes, it does.

kxepal commented on ansible/roles/ubuntu-worker/tasks/rebar.yml:
```diff
@@ -0,0 +1,37 @@
+
+- name: download rebar
+  get_url:
+    url: https://github.com/rebar/rebar/archive/{{ rebar_version }}.tar.gz
```
No need in rebar: ./configure script [takes care](https://github.com/apache/couchdb/blob/master/configure#L652-L667) about it. While you stay with `make *` commands, it will be used and right version.

basti1302 replied:
Oh, that's good to know. I'll remove the whole rebar stuff. Funny though, both https://github.com/apache/couchdb/blob/master/Dockerfile and https://github.com/klaemo/docker-couchdb/blob/2.0-dev/Dockerfile install rebar separately.

kxepal replied:
That is need of old past days. Not saying that this is wrong and wont work, but if you can make things simpler - why not?(:

kxepal commented on readme.markdown:
```diff
@@ -3,44 +3,80 @@ CouchDB CI Setup
-*Remark: Throughout this repository we use the terms "master"/"worker" for the Jenkins build machines, whereas the Jenkins documentation uses the terms "master"/"slave".*
+### Open questions
+
+* AFAIK Erlang 14 support will be dropped soon-ish, so I'm not sure if it is worth the effort to do anything for that.
```
Erlang R14 support is already dropped with no return. So R16B03-1+.

kxepal commented on readme.markdown:
```diff
@@ -3,44 +3,80 @@ CouchDB CI Setup
+### Open questions
+
+* AFAIK Erlang 14 support will be dropped soon-ish, so I'm not sure if it is worth the effort to do anything for that.
+* Which 18.x Erlang version is to be used? I heard someone saying 18.0 once, but that was before 18.1 and 18.2 were available, so I guess it makes more sense to always use the latest 18.x to see if changes in Erlang 18 breaks CouchDB.
```
Imho, need to ensure that we run correctly with the latest and system default if it matches the requirement.

rnewson replied:
while it's important to test against each OS's stock erlang install (where available), I think it's important to test various erlang versions too, we need to know that our software works on previous and future versions (relative to what's in any stable OS release). Take travis as an example, we can list any (or all) erlang versions there. The main deficiency of travis is that we can't also have an OS axis to that matrix.

rnewson replied:
so perhaps there's two parts to this 1) confirming couchdb tests run with the standard dependencies from each of the various important OS/distro combinations and 2) confirming couchdb tests run against many versions of important dependencies like erlang itself, even if the OS in those cases remains the same.

kxepal commented:
Hi @basti1302!

Great to see the progress here! (: Regarding the functional part of your work, you need to add [license headers](http://apache.org/legal/src-headers.html#headers) for your source files. ASF rules. If you have some work made by third party - that would be need to notice as well.

basti1302 commented:
Hey @kxepal, 

thanks for your input. I'll update the PR according to your suggestions.

rnewson commented:
One of the requirements of a CI environment is to ensure we don't break support on some platform or other. Correct me if I'm wrong but Docker is Linux only? I'd like to see a CI environment that works, at least in theory, on BSD and other platforms.

I know @janl has started work on a Vagrant-based system over at https://github.com/janl/couchdb-ci-build-vms and I've contributed Debian and FreeBSD setup.

basti1302 commented:
Actually, I started out building Vagrant VMs provisioned with Ansible. Docker was what ASF Infra suggested to get the various Linux distros up and running quickly. For Mac, Windows and possibly FreeBSD we need another solution.

That's also the reason why the main work is done in Ansible. You can use the Ansible scripts to provision Docker containers, Vagrant VMs, bare metal, anything you like.

So the plan is to use the same set of Ansible scripts for all situations but use them to provsion Docker containers for all the Linux flavors and do "something else" for the other OSes.

rnewson commented:
thanks for the explanation, that sounds good to me then.

rnewson commented on readme.markdown:
```diff
@@ -3,44 +3,80 @@ CouchDB CI Setup
+* There is no CentOS/RHEL there, shouldn't it be added?
+* Do we run a CouchDB build on all combinations on each commit? This would probably be too much for the ASF Infra build systems. Do we build them once a day? We need to find a good balance between early feedback and resource consumption here.
+* Do we even want to build the master branch or some other branch/tag? I guess the master branch would be most interesting for now, but not entirely sure. Also, it might make sense to make the branch/tag parameterizable so we could also use this to create releases from a specific tag etc.
+* What exactly do we do in each Jenkins build? Just build CouchDB? Also build docs? Start CouchDB? Run some test suite?
```
build couchdb, build docs, run all test suites.

The 'build docs' step is very onerous, taking quite a while. If possible, we'd avoid building docs again if they haven't changed since last time, though that might be difficult to achieve.

basti1302 commented:
Update: I think I addressed all reviewer's comment that were immediately actionable. 

In particular: 
- Add license headers to all files
- Added dependencies required for `make check` (erlang-eunit, shunit2, ./configure with --with-curl)
- Execute target `make all check dist` instead of only `make`
- Remove manual steps to download and compile rebar

I'd say it is ready to merge for now. I would like to approach the INFRA team and ask them to set up the first CouchDB build based on this Docker container. Expansion to other Erlang versions and OSes can continue in parallel, but that's for another PR.

basti1302 commented:
Oh, I nearly forgot. Thank you all very much for the fast feedback and help. Much appreciated. <3
